From 1d63a2be9dd396dd11b5dcf6c9e95f4f408d08cd Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 26 Aug 2022 14:20:35 +0300
Subject: [PATCH] [SW-86869] gaudi3: halt rom in case of fuse init failure

In previous platforms, whenever fuse init failed the default value of 0
was assigned to syscfg fuse value which essentially means disabling all
security. This also mean that boot flow would have continued with images
signed with revoked keys on secured system. Since there is no correct
default value of syscfg fuse, this patch halt rom execution in case of
fuse init failure. This was recommended as part of secured code review
and also aligns the implementation with HW FSM.

Change-Id: I89345cde2f04fce822386854a7ad4b8a1607a55e
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index cfaaae5..2246c3d 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -498,12 +498,31 @@
 	return !!(arcboot_rom_global.config_fuse & EFUSE_SYSCFG_SKU_MASK);
 }
 
+void arcboot_rom_halt(void)
+{
+	while (1)
+		wfe();
+}
+
 void arcboot_rom_fuse(void)
 {
 	uint32_t val;
 
-	if (fuse_init() == 0)
+	/*
+	 * In case of fuse init failure, recommendation is
+	 * to halt the system as there is no right default value
+	 * which can be set or used for efuse if unable to init
+	 * efuse successfully. Secondly, HW FSM also performs init
+	 * of fuse in similar fashion and if failed enter IDLE state
+	 * and doesn't proceed with boot. similar approach is used
+	 * here in case of efuse init failure.
+	 */
+	if (fuse_init() == 0) {
 		arcboot_rom_set_efuse_read();
+	} else {
+		arcboot_rom_status_update(PROGRESS_INIT_EFUSE_INIT_FAIL);
+		arcboot_rom_halt();
+	}
 
 	if (fuse_read(EFUSE_SYSCFG_OFFSET, &val) != 0)
 		val = EFUSE_SYSCFG_DEFAULT;
diff --git a/arch/gaudi3/arc/dev_common.h b/arch/gaudi3/arc/dev_common.h
index 27ea0a5..b8c355b 100644
--- a/arch/gaudi3/arc/dev_common.h
+++ b/arch/gaudi3/arc/dev_common.h
@@ -31,3 +31,4 @@
 void disable_spi_gw(void);
 uint64_t get_pcie_pll_freq(enum pll_div_out div);
 enum dma_dest_type arcboot_rom_get_dma_dest_type(uint64_t addr);
+void arcboot_rom_halt(void);
diff --git a/arch/gaudi3/arc/progress_bits.h b/arch/gaudi3/arc/progress_bits.h
index 64cc6c2..b14f90e 100644
--- a/arch/gaudi3/arc/progress_bits.h
+++ b/arch/gaudi3/arc/progress_bits.h
@@ -23,128 +23,129 @@
 	PROGRESS_INIT_SPI_BOOT_DESC_INFO_GET_FAIL = 16,
 	PROGRESS_INIT_SPI_BOOT_DESC_DATA_GET_FAIL = 17,
 	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 18,
-	PROGRESS_INIT_EFUSE_SPI = 19,
-	PROGRESS_INIT_EFUSE_SPI_SLV_TIMEDOUT = 20,
-	PROGRESS_REVOKE_LOC = 21,
-	PROGRESS_REVOKE_LOC_0 = 22,
-	PROGRESS_REVOKE_LOC_1 = 23,
-	PROGRESS_KEY_LOC = 24,
-	PROGRESS_KEY_LOC_0 = 25,
-	PROGRESS_KEY_LOC_1 = 26,
-	PROGRESS_VER_LOC = 27,
-	PROGRESS_VER_LOC_0 = 28,
-	PROGRESS_VER_LOC_1 = 29,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_START = 30,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_VERSION_FAIL = 31,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_REVOKED = 32,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 33,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 34,
-	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 35,
-	PROGRESS_BRINGUP_PCI_PRI = 36,
-	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 37,
-	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 38,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 39,
-	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 40,
-	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 41,
-	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 42,
-	PROGRESS_BRINGUP_PCI_PRI_DMA = 43,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 44,
-	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 45,
-	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 46,
-	PROGRESS_BRINGUP_PCI_SEC = 47,
-	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 48,
-	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 49,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 50,
-	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 51,
-	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 52,
-	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 53,
-	PROGRESS_BRINGUP_PCI_SEC_DMA = 54,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 55,
-	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 56,
-	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 57,
-	PROGRESS_ARCBOOT_IN_RECOVERY = 58,
-	PROGRESS_ARCBOOT_ROM_TPM = 59,
-	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 60,
-	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 61,
-	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 62,
-	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 63,
-	PROGRESS_BRINGUP_TPM_SPI = 64,
-	PROGRESS_BRINGUP_TPM_SPI_PRI = 65,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 66,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 67,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 68,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 69,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 70,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 71,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 72,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 73,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 74,
-	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 75,
-	PROGRESS_BRINGUP_TPM_SPI_SEC = 76,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 77,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 78,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 79,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 80,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 81,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 82,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 83,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 84,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 85,
-	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 86,
-	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 87,
-	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 88,
-	PROGRESS_PPBOOT_START = 89,
-	PROGRESS_PPBOOT_FINISH = 90,
-	PROGRESS_BRINGUP_RAM_PRI = 91,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 92,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 93,
-	PROGRESS_BRINGUP_RAM_SEC = 94,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 95,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 96,
-	PROGRESS_ARCBOOT_RAM_START = 97,
-	PROGRESS_ARCBOOT_RECOVERY = 98,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 99,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 100,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 101,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 102,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 103,
-	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 104,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 105,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 106,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 107,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 108,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 109,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 110,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 111,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 112,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 113,
-	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 114,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 115,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 116,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 117,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 118,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 119,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 120,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 121,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 122,
-	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 123,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 124,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 125,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 126,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 127,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 128,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 129,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 130,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 131,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 132,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 133,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 134,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 135,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 136,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 137,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 138,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 139,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 140,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 141,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 142,
+	PROGRESS_INIT_EFUSE_INIT_FAIL = 19,
+	PROGRESS_INIT_EFUSE_SPI = 20,
+	PROGRESS_INIT_EFUSE_SPI_SLV_TIMEDOUT = 21,
+	PROGRESS_REVOKE_LOC = 22,
+	PROGRESS_REVOKE_LOC_0 = 23,
+	PROGRESS_REVOKE_LOC_1 = 24,
+	PROGRESS_KEY_LOC = 25,
+	PROGRESS_KEY_LOC_0 = 26,
+	PROGRESS_KEY_LOC_1 = 27,
+	PROGRESS_VER_LOC = 28,
+	PROGRESS_VER_LOC_0 = 29,
+	PROGRESS_VER_LOC_1 = 30,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_START = 31,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_VERSION_FAIL = 32,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_REVOKED = 33,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 34,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_AUTH_FAIL = 35,
+	PROGRESS_SPI_BOOT_DESC_VERIFY_SUCCESS = 36,
+	PROGRESS_BRINGUP_PCI_PRI = 37,
+	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 38,
+	PROGRESS_BRINGUP_PCI_PRI_CRC_FAIL = 39,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_DEST_ADDR = 40,
+	PROGRESS_BRINGUP_PCI_PRI_INVALID_IMG_SIZE = 41,
+	PROGRESS_BRINGUP_PCI_PRI_UNALIGNED_IMG_SIZE = 42,
+	PROGRESS_BRINGUP_PCI_PRI_DMA_ERROR = 43,
+	PROGRESS_BRINGUP_PCI_PRI_DMA = 44,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_REVOKED = 45,
+	PROGRESS_BRINGUP_PCI_PRI_KEY_AUTH_FAIL = 46,
+	PROGRESS_BRINGUP_PCI_PRI_SECURITY = 47,
+	PROGRESS_BRINGUP_PCI_SEC = 48,
+	PROGRESS_BRINGUP_PCI_SEC_TOO_BIG = 49,
+	PROGRESS_BRINGUP_PCI_SEC_CRC_FAIL = 50,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_DEST_ADDR = 51,
+	PROGRESS_BRINGUP_PCI_SEC_INVALID_IMG_SIZE = 52,
+	PROGRESS_BRINGUP_PCI_SEC_UNALIGNED_IMG_SIZE = 53,
+	PROGRESS_BRINGUP_PCI_SEC_DMA_ERROR = 54,
+	PROGRESS_BRINGUP_PCI_SEC_DMA = 55,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_REVOKED = 56,
+	PROGRESS_BRINGUP_PCI_SEC_KEY_AUTH_FAIL = 57,
+	PROGRESS_BRINGUP_PCI_SEC_SECURITY = 58,
+	PROGRESS_ARCBOOT_IN_RECOVERY = 59,
+	PROGRESS_ARCBOOT_ROM_TPM = 60,
+	PROGRESS_ARCBOOT_ROM_TPM_INIT_FAIL = 61,
+	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 62,
+	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 63,
+	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 64,
+	PROGRESS_BRINGUP_TPM_SPI = 65,
+	PROGRESS_BRINGUP_TPM_SPI_PRI = 66,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 67,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 68,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 69,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_INVALID_IMG_SIZE = 70,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 71,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 72,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 73,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 74,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 75,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 76,
+	PROGRESS_BRINGUP_TPM_SPI_SEC = 77,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 78,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 79,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 80,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_INVALID_IMG_SIZE = 81,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 82,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 83,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 84,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 85,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 86,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 87,
+	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 88,
+	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 89,
+	PROGRESS_PPBOOT_START = 90,
+	PROGRESS_PPBOOT_FINISH = 91,
+	PROGRESS_BRINGUP_RAM_PRI = 92,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 93,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 94,
+	PROGRESS_BRINGUP_RAM_SEC = 95,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 96,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 97,
+	PROGRESS_ARCBOOT_RAM_START = 98,
+	PROGRESS_ARCBOOT_RECOVERY = 99,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 100,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 101,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 102,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 103,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_SRAM_INIT_TIMEDOUT = 104,
+	PROGRESS_ARCBOOT_ROM_PCIE_PHY_PARA_STS_CLR_TIMEDOUT = 105,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 106,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 107,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 108,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 109,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 110,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 111,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 112,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 113,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 114,
+	PROGRESS_INIT_EROM_BOOT_DESC_EXCEEDS_MAX_SIZE = 115,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 116,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 117,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 118,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_SIZE = 119,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 120,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 121,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 122,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 123,
+	PROGRESS_INIT_EROM_BOOT_DESC_INVALID_IMG_OFFSET = 124,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 125,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 126,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 127,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 128,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 129,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 130,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 131,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 132,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 133,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 134,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 135,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 136,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_INVALID_IMG_SIZE = 137,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 138,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 139,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 140,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 141,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 142,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 143,
 };
diff --git a/arch/gaudi3/arc/recovery.c b/arch/gaudi3/arc/recovery.c
index f8ee5e7..68afffa 100644
--- a/arch/gaudi3/arc/recovery.c
+++ b/arch/gaudi3/arc/recovery.c
@@ -248,6 +248,5 @@
 	 * download and verification error.
 	 * Stay here (sleep) and do not return
 	 */
-	while (1)
-		wfe();
+	arcboot_rom_halt();
 }
